### ISSUE TITLE
Remove read-to-read barrier warning for barriers with src=VK_ACCESS_TRANSFER_READ_BIT and dst=VK_ACCESS_SHADER_READ_BIT

### ISSUE THREAD

**Issue by Izhido** [KhronosGroup/Vulkan-ValidationLayers#9507]
This warning appears when enabling NVIDIA-side best practices layer.
As far as I know, no other layers ever complain about this.

I can understand that read-to-read barriers do not make sense... **unless** they are performed for different stages in the pipeline. To warn blindly that it is a read-to-read barrier, without analyzing the rest of the data, is not really useful as a validation or performance warning.

Request to remove the warning, or to modify the validation to analyze also the remaining barrier data before issuing a warning.

**Comment by spencer-lunarg** [comment]
@Izhido can you post the warning message it post (so we can make sure we are agreeing which `LogWarning` in the code this is referring to)

**Comment by Izhido** [comment]
Sure thing. The message is as follows:

`Validation Performance Warning: [ BestPractices-PipelineBarrier-readToReadBarrier ] Object 0: handle = 0x2c7e28e53e0, type = VK_OBJECT_TYPE_COMMAND_BUFFER; | MessageID = 0x2f637ff | vkCmdPipelineBarrier(): [AMD] [NVIDIA] Don't issue read-to-read barriers. Get the resource in the right state the first time you use it.`

This message appears after the following code is executed, at the end of a FillMipmapped() function in my code:

```
		barrier.srcAccessMask = VK_ACCESS_TRANSFER_READ_BIT;
		barrier.dstAccessMask = VK_ACCESS_SHADER_READ_BIT;
		barrier.oldLayout = VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL;
		barrier.newLayout = VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL;
		barrier.subresourceRange.baseMipLevel = mips - 1;
		barrier.subresourceRange.levelCount = remaining + 1;
		vkCmdPipelineBarrier(buffer.commandBuffer, VK_PIPELINE_STAGE_TRANSFER_BIT, VK_PIPELINE_STAGE_FRAGMENT_SHADER_BIT, 0, 0, nullptr, 0, nullptr, 1, &barrier);
```

**Comment by spencer-lunarg** [comment]
I agree that `BestPractices-PipelineBarrier-readToReadBarrier` has just been "left" there since the AMD/NVIDIA engineers added it.

@artem-lunarg thoughts on the warning

**Comment by artem-lunarg** [comment]
**General Vulkan**
At first, a general (not vendor specific) notes. If we are talking only about memory barriers (without layout transition) then READ-to-READ barrier is not needed. Two reads can perfectly run in parallel witout causing any race condition (does not matter if it's the same stage or different). Race condition can only happen when there is at least one write. 

Going to the case when READ-to-READ *image* barrier defines a **layout transition**, this operation is not no-op (does actual work) due to layout transition operation. So it **must be specified**.

Please note, if we create a barrier where *source* stage is a READ operation then memory barrier is not needed and *execution* dependency is enough! That's because *destination* stage waits for when execution of the previous READ operation is finished, but does not need to ensure *availability* of that READ (needed only for write operations, which needs to be propagated over the memory hierachy, and read operation is finished as soon as corresponding instruction finished execution). To define an execution dependency the barrier should specify srcStage/dstStage stage but set srcAccess/dstAccess to zero.

**Specific vendors**
Back to the actual GPU hardware. How layout transition operation is implemented it's up to the GPU vendor, it could be a no-op (the same memory representation is used in all contexts) or image representation is  changed only for specific pairs of Vulkan's IMAGE_LAYOUT constants but not for the others, which reflects chip/driver design. In this case **AMD/NVIDIA recommends what is best for theirs hardware**.  _I don't see any issues with this recommentation._ It's vendor-specific recommendation, it's fine to ignore it (it's not violation of the vulkan spec, but can get additional performance benefits on specific hardware).

Regarding implementation details of this vendor-specific validation. In addition to checking that barrier is READ-to-READ, **it also checks old and new layouts**.  If specific layout pairs are detected for which layout transition is not performed this generates vendor specific recommendation message. In this specific case the valdiation checks if both layouts are one of the following read-only layouts (VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL or VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL or 
VK_IMAGE_LAYOUT_TRANSFER_SRC_OPTIMAL). It means that driver uses the same image representation for these layouts.

It's a design choice how to take this into account. Personally I consider this as a hint to fine tune performance when multi-vendor Vulkan code already works correclty and there is a *measurable* performance improvement from implementing vendor-specific pass that outweights the added complexity. In some cases software is strongly oriented to specific vendors (based on data center configuration, etc).
